koku/api/report/ocp/serializers.py

```diff
@@ -40,11 +44,15 @@ class OCPOrderBySerializer(OrderSerializer):
 
     def to_internal_value(self, data):
         """Send to internal value."""
-        if not isinstance(data, dict):
-            return super().to_internal_value(data)
-        for serializer_key, internal_key in DISTRIBUTED_COST_INTERNAL.items():
-            if serializer_key in data.keys():
-                data[internal_key] = data.pop(serializer_key)
+        try:
+            keys = data.keys()
+        except AttributeError as exc:
+            LOG.error(f"Unexpected data encountered during serialization: {exc}")
```

Comment:
What do we gain by logging this error? The user supplied invalid input. All this error will do is spam our alert channel.

This was also why I simply checked if it was a dict. If it is not a dict, then we don't need to modify the internal value at all.

Reply:
Another idea is to do this:
```
        data = super().to_internal_value(data)
        for serializer_key, internal_key in DISTRIBUTED_COST_INTERNAL.items():
            if serializer_key in data.keys():
                data[internal_key] = data.pop(serializer_key)
        return data
```
use super first to validate the data structure, then modify what we have after we know it is valid.

Reply:
The logging there was to avoid swallowing the exception details. But you're right this would be way too noisy.

Reply:
I was trying to avoid the "super sandwich". That looked funny even though it worked.

Doesn't the data need to be run through `super().to_internal_value(data)` _after_ the keys in `data` are potentially changed?

Reply:
Hmm, honestly, this might be the wrong function entirely. We should probably be using `to_representation` here instead of `to_internal_value`:

https://www.django-rest-framework.org/api-guide/serializers/#overriding-serialization-and-deserialization-behavior

Need to think about this one more.

Reply:
I clearly got very distracted here. Anyway, I don't think a try/except is necessarily correct here. These methods are meant to validate data, and I don't think a "ask for forgiveness" pattern is really doing that. I could be down for adding nesting if you're not a fan of the super sandwich:
```
        if isinstance(data, dict):
            for serializer_key, internal_key in DISTRIBUTED_COST_INTERNAL.items():
                if serializer_key in data.keys():
                    data[internal_key] = data.pop(serializer_key)
        return super().to_internal_value(data)
```

Reply:
Looking at what's happening inside `to_internal_value()` it looks like we get primitive validation of the contents of a given field to make sure it contains values in `OrderSerializer.ORDER_CHOICES`.

I think you're correct: an instance check makes the most sense in validation code. I'll go back to basically what you had initially.

Reply:
Thanks for scrutinizing this. I learned some things about DRF and like the solution we ended up with.